### PR TITLE
GH-625 Adds form input types to marshaller/HipieDDL.js

### PR DESCRIPTION
**Pull request by mlzummo**
@GordonSmith @dtsnell4 @jbrundage 

Please Review
@jbrundage there is multiple domnodes in the input for checkbox, this is just the best solution otherwise i was going to create a "select" widget which just did the same thing anything anyways

Adds form input types to marshaller
Fixes GH-625

Signed-off-by: Mathew Zummo <mzummo@gmail.com>

### CODE REVIEW

File: src/api/IInput.js [hunk 1]
```diff
@@ -8,12 +8,15 @@
...
         Widget.call(this);
+
+        this.checkedVals = [];
     }
     IInput.prototype = Object.create(Widget.prototype);
 
     IInput.prototype.publish("name", "", "string", "HTML name for the input");
     IInput.prototype.publish("label", "", "string", "Descriptive label");
-    IInput.prototype.publish("value", "", "string", "Input type");
+    //IInput.prototype.publish("value", "", "string", "Input type");
```
Comment (mlzummo) [inline src/api/IInput.js]: remove comment

File: src/api/IInput.js [hunk 2]
```diff
@@ -8,12 +8,15 @@
 }(this, function (Widget) {
     function IInput() {
         Widget.call(this);
+
+        this.checkedVals = [];
```
Comment (mlzummo) [inline src/api/IInput.js]: not used remove

File: src/marshaller/HipieDDL.js [hunk 3]
```diff
@@ -571,17 +571,67 @@
                     });
                     break;
                 case "FORM":
-                    this.loadWidgets(["src/form/Form", "src/form/Input"], function (widget, widgetClasses) {
+                    this.loadWidgets(["src/form/Form", "src/form/Input", "src/layout/Grid"], function (widget, widgetClasses) {
```
Comment (mlzummo) [inline src/marshaller/HipieDDL.js]: remove grid

File: src/form/Input.js [hunk 4]
```diff
@@ -35,67 +37,102 @@
...
+                this._inputElement[0] = element.append("textarea");
                 break;
             case "button":
-                this._inputElement = element.append("button");
+                this._inputElement[0] = element.append("button");
+                break;
+            case "radio":
+                /* falls through */
+            case "checkbox":
+                var checkboxContainer = element.append("ul");
```
Comment (mlzummo) [inline src/form/Input.js]: need to class this

**Comment by mlzummo** [comment]
@drealeed This is what I have so far to resolve this issue:

rawgit.com/mlzummo/Visualization/GH_625_FORM_INP_MARSHALLER_UPDATE/demos/marshaller/index.html?Hostname=10.173.147.1&Port=8010&Protocol=http&Wuid=W20150814-181122&ResultName=leeddx_issue_780_formwidget_Comp_Ins002_DDL

**Comment by mlzummo** [comment]
fixes-736 as well (add to commit msg)

File: src/form/Form.js [hunk 5]
```diff
@@ -213,6 +214,7 @@
 
                 var inputWidgetArray = inputWidget instanceof WidgetArray ? inputWidget.content() : [inputWidget];
                 inputWidgetArray.forEach(function (inputWidget, idx) {
+
```
Comment (dtsnell4) [inline src/form/Form.js]: space

File: src/form/Input.js [hunk 6]
```diff
@@ -11,6 +11,8 @@
         IInput.call(this);
 
         this._tag = "div";
+
```
Comment (dtsnell4) [inline src/form/Input.js]: space

File: src/form/Form.js [hunk 7]
```diff
@@ -39,7 +39,7 @@
                     .name("select-test")
                     .label("Select Test")
                     .type("select")
-                    .selectOptions(["A","B","C"])
+                    .selectOptions([["AAA"],["BBB"],["CCC"]])
```
Comment (GordonSmith) [inline src/form/Form.js]: Whay is this an array of arrays? what does the following do:
```
 .selectOptions([["AAA", "AAA2"],["BBB", "BBB2"],["CCC", "CCC2"]])
```

**Comment by mlzummo** [comment]
@GordonSmith 
select options can contain what will be displayed and value pairs now ex:

.selectOptions([["FLORIDA","FL"],["colorado", "co"],["California", "CA"]])

so your example would show an option with the display value of AAA and with the value actully set to  AAA2

the login for that is here: https://github.com/hpcc-systems/Visualization/pull/916/files#diff-76a4bf71f46d512fa2c6923c1663d91aR144

**Comment by GordonSmith** [comment]
Needs rebase

File: src/form/Form.js [hunk 8]
```diff
@@ -78,6 +78,7 @@
         if (this.validate()) {
             isValid = this.checkValidation();
         }
+        console.log(isValid ? this.values() : null); // for testing putput (notice array conv to comma seperated string)
```
Comment (GordonSmith) [inline src/form/Form.js]: remove console.log

**Comment by GordonSmith** [comment]
http://rawgit.com/mlzummo/Visualization/GH_625_FORM_INP_MARSHALLER_UPDATE/demos/dermatology.html?src/form/Form
Is not working (the checkbox input field)

**Comment by mlzummo** [comment]
@GordonSmith Fixed

**Comment by GordonSmith** [comment]
@jbrundage @dtsnell4 Please review

File: src/form/Input.js [hunk 9]
```diff
@@ -27,93 +28,128 @@
...
+                        break;
+                }
+                w.change(w);
+            });
         });
-        if(context._inputElement2 !== undefined){
-            context._inputElement2.on("change",function(w){
-                context.value(context._inputElement2.property("value"));
-                context._inputElement.property("value",context._inputElement2.property("value"));
+        if(context._inputElement[1] !== undefined){
```
Comment (jbrundage) [inline src/form/Input.js]: Looks like this will have adverse effects on types that arent html-color

**Comment by jbrundage** [comment]
Looks good to me

**Comment by dtsnell4** [comment]
No new comments, looks good

**Pull request merged**